### PR TITLE
Updating methods to follow LLVM 17 APIs

### PR DESCRIPTION
**Pull request by advikkabra**
Work in progress to get all integration tests working.

### CODE REVIEW

**Comment by advikkabra** [comment]
This has somehow caused `print_arr_03` to fail, I will fix that then move to the next test.

**Comment by certik** [comment]
I recommend to fix one thing per PR, so once you figure out how to make all tests pass, we can merge it.

**Comment by advikkabra** [comment]
After this PR, I'll start splitting it up.

**Comment by certik** [comment]
Ok. Please ping me once tests pass.

**Comment by certik** [comment]
I made it a Draft -- once CI passes and this is ready, click on "Ready for review".

File: src/libasr/codegen/llvm_array_utils.cpp [hunk 1]
```diff
@@ -666,8 +671,9 @@ namespace LCompilers {
                     tmp = llvm_utils->create_ptr_gep2(type, array, idx);
                 }
             } else {
-                idx = cmo_convertor_single_element(array, m_args, n_args, check_for_bounds);
-                llvm::Value* full_array = get_pointer_to_data(array);
+                llvm::Type* array_type = get_array_type(asr_type, type);
+                idx = cmo_convertor_single_element(array_type, array, m_args, n_args, check_for_bounds);
+                llvm::Value* full_array = get_pointer_to_data(array_type, array);
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/llvm_array_utils.cpp:676]: The type might already be there in the ptr_type map.

File: src/libasr/codegen/llvm_array_utils.cpp [hunk 2]
```diff
@@ -599,10 +604,10 @@ namespace LCompilers {
         // void check_single_element(llvm::Value* curr_idx, llvm::Value* arr) {
         // }
 
-        llvm::Value* SimpleCMODescriptor::cmo_convertor_single_element(
+        llvm::Value* SimpleCMODescriptor::cmo_convertor_single_element(llvm::Type *type,
             llvm::Value* arr, std::vector<llvm::Value*>& m_args,
             int n_args, bool check_for_bounds) {
-            llvm::Value* dim_des_arr_ptr = llvm_utils->CreateLoad2(dim_des->getPointerTo(), llvm_utils->create_gep(arr, 2));
+            llvm::Value* dim_des_arr_ptr = get_pointer_to_dimension_descriptor_array(type, arr);
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/llvm_array_utils.cpp:610]: Same, the type might be in the map.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 3]
```diff
@@ -9877,24 +9885,26 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
         llvm::Value* llvm_arg1 = tmp;
         visit_expr_wrapper(x.m_dim, true);
         llvm::Value* dim_val = tmp;
 
-        ASR::ttype_t* x_mv_type = ASRUtils::expr_type(x.m_v);
         ASR::array_physical_typeType physical_type = ASRUtils::extract_physical_type(x_mv_type);
         switch( physical_type ) {
             case ASR::array_physical_typeType::DescriptorArray: {
-                llvm::Value* dim_des_val = arr_descr->get_pointer_to_dimension_descriptor_array(llvm_arg1);
+                llvm::Value* dim_des_val = arr_descr->get_pointer_to_dimension_descriptor_array(array_type, llvm_arg1);
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_llvm.cpp:9907]: Same, type is there in the map.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 4]
```diff
@@ -5263,7 +5270,8 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
                     ASRUtils::type_get_past_pointer(m_type)), module.get());
             llvm::AllocaInst *target = llvm_utils->CreateAlloca(
                 target_type, nullptr, "array_descriptor");
-            builder->CreateStore(llvm_utils->create_ptr_gep(
+            builder->CreateStore(llvm_utils->create_ptr_gep2(
+                data_type,
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_llvm.cpp:5274]: ..

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 5]
```diff
@@ -9771,7 +9779,7 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
                     LLVM::is_llvm_pointer(*ASRUtils::expr_type(m_v));
         visit_expr_wrapper(m_v);
         ptr_loads = ptr_loads_copy;
-        bool is_pointer_array = (tmp->getType()->getNumContainedTypes() > 0) && (tmp->getType()->getContainedType(0)->isPointerTy());
+        bool is_pointer_array = ((tmp->getType()->getNumContainedTypes() > 0) && (tmp->getType()->getContainedType(0)->isPointerTy()));
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_llvm.cpp:9782]: Revert this?

**Review by Thirumalai-Shaktivel — COMMENTED** [review]
Great, the changes seems to work and pass all the tests. 

Some suggestions, most of the places the types are available in the map. I think all the changes in `llvm_array_utils.h` can be reverted. Let's keep the changes minimal.

**Review by Thirumalai-Shaktivel — APPROVED** [review]

**Pull request merged**